### PR TITLE
build: pass ignore file to buildah to consider empty ignore files

### LINKED ISSUE #25147: When --ignorefile is used and it's empty, it causes error with podman build
### Issue Description

When trying to build a podman container, I noticed that a file wasn't being included that I expected to be there.  Upon closer investigation, I noticed that a .dockerignore was present in the root directory causing the file not to be copied.

I tried overriding with `--ignorefile /dev/null` and continued to get the error.

I narrowed it down to what I believe to be the root cause: an empty ignorefile seems to fallback to the default docker ignorefile.

I'd expect if the `--ignorefile` flag is explicitly set to an empty file, then it should not load the original ignorefile.

### Steps to reproduce the issue

```
abraithwaite at gibson in ~/Projects/github.com/runreveal/example
23:10:39 $ tree .
.
├── Dockerfile
├── dist
│   ├── altignore.txt
│   ├── asdf1.txt
│   └── asdf2.txt
├── test1.txt
└── test2.txt

2 directories, 6 files

abraithwaite at gibson in ~/Projects/github.com/runreveal/example
23:10:43 $ cat dist/altignore.txt


abraithwaite at gibson in ~/Projects/github.com/runreveal/example
23:14:56 $ cat .dockerignore
test2.txt
**/dist

abraithwaite at gibson in ~/Projects/github.com/runreveal/example
23:10:47 $ cat Dockerfile
FROM alpine:latest
COPY dist/asdf* /tmp/

abraithwaite at gibson in ~/Projects/github.com/runreveal/example
23:11:25 $ podman build --no-cache -t testignore --ignorefile dist/altignore.txt .
STEP 1/2: FROM alpine:latest
STEP 2/2: COPY dist/asdf* /tmp/
Error: building at STEP "COPY dist/asdf* /tmp/": checking on sources under "/var/tmp/libpod_builder1578895609/build": Rel: can't make  relative to /var/tmp/libpod_builder1578895609/build; copier: stat: ["/dist/asdf*"]: no such file or directory
```

### Describe the results you received

A somewhat obscure error is thrown and I was left wondering what the heck was happening.

```
Error: building at STEP "COPY dist/asdf* /tmp/": checking on sources under "/var/tmp/libpod_builder1578895609/build": Rel: can't make  relative to /var/tmp/libpod_builder1578895609/build; copier: stat: ["/dist/asdf*"]: no such file or directory
```

### Describe the results you expected

I expect that: when passing in an empty ignore file, the original `.dockerignore` or `.containerignore` is overridden and not parsed.

As such, an empty ignore file shouldn't ignore anything.

The use case is 

### podman info output

```yaml
host:
  arch: arm64
  buildahVersion: 1.35.3
  cgroupControllers:
  - cpuset
  - cpu
  - io
  - memory
  - pids
  - rdma
  - misc
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.1.10-1.fc39.aarch64
    path: /usr/bin/conmon
    version: 'conmon version 2.1.10, commit: '
  cpuUtilization:
    idlePercent: 88.47
    systemPercent: 2.83
    userPercent: 8.7
  cpus: 4
  databaseBackend: sqlite
  distribution:
    distribution: fedora
    variant: coreos
    version: "39"
  eventLogger: journald
  freeLocks: 1597
  hostname: localhost.localdomain
  idMappings:
    gidmap: null
    uidmap: null
  kernel: 6.8.4-200.fc39.aarch64
  linkmode: dynamic
  logDriver: journald
  memFree: 2460356608
  memTotal: 8301178880
  networkBackend: netavark
  networkBackendInfo:
    backend: netavark
    dns:
      package: aardvark-dns-1.10.0-1.fc39.aarch64
      path: /usr/libexec/podman/aardvark-dns
      version: aardvark-dns 1.10.0
    package: netavark-1.10.3-1.fc39.aarch64
    path: /usr/libexec/podman/netavark
    version: netavark 1.10.3
  ociRuntime:
    name: crun
    package: crun-1.14.4-1.fc39.aarch64
    path: /usr/bin/crun
    version: |-
      crun version 1.14.4
      commit: a220ca661ce078f2c37b38c92e66cf66c012d9c1
      rundir: /run/crun
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +LIBKRUN +WASM:wasmedge +YAJL
  os: linux
  pasta:
    executable: /usr/bin/pasta
    package: passt-0^20240405.g954589b-1.fc39.aarch64
    version: |
      pasta 0^20240405.g954589b-1.fc39.aarch64-pasta
      Copyright Red Hat
      GNU General Public License, version 2 or later
        <https://www.gnu.org/licenses/old-licenses/gpl-2.0.html>
      This is free software: you are free to change and redistribute it.
      There is NO WARRANTY, to the extent permitted by law.
  remoteSocket:
    exists: true
    path: /run/podman/podman.sock
  rootlessNetworkCmd: ""
  security:
    apparmorEnabled: false
    capabilities: CAP_CHOWN,CAP_DAC_OVERRIDE,CAP_FOWNER,CAP_FSETID,CAP_KILL,CAP_NET_BIND_SERVICE,CAP_SETFCAP,CAP_SETGID,CAP_SETPCAP,CAP_SETUID,CAP_SYS_CHROOT
    rootless: false
    seccompEnabled: true
    seccompProfilePath: /usr/share/containers/seccomp.json
    selinuxEnabled: true
  serviceIsRemote: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.2.2-1.fc39.aarch64
    version: |-
      slirp4netns version 1.2.2
      commit: 0ee2d87523e906518d34a6b423271e4826f71faf
      libslirp: 4.7.0
      SLIRP_CONFIG_VERSION_MAX: 4
      libseccomp: 2.5.3
  swapFree: 0
  swapTotal: 0
  uptime: 3h 26m 28.00s (Approximately 0.12 days)
  variant: v8
plugins:
  authorization: null
  log:
  - k8s-file
  - none
  - passthrough
  - journald
  network:
  - bridge
  - macvlan
  - ipvlan
  volume:
  - local
registries:
  search:
  - docker.io
store:
  configFile: /usr/share/containers/storage.conf
  containerStore:
    number: 17
    paused: 0
    running: 5
    stopped: 12
  graphDriverName: overlay
  graphOptions:
    overlay.imagestore: /usr/lib/containers/storage
    overlay.mountopt: nodev,metacopy=on
  graphRoot: /var/lib/containers/storage
  graphRootAllocated: 106769133568
  graphRootUsed: 60304642048
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Supports shifting: "true"
    Supports volatile: "true"
    Using metacopy: "true"
  imageCopyTmpDir: /var/tmp
  imageStore:
    number: 199
  runRoot: /run/containers/storage
  transientStore: false
  volumePath: /var/lib/containers/storage/volumes
version:
  APIVersion: 5.0.2
  Built: 1713312000
  BuiltTime: Tue Apr 16 17:00:00 2024
  GitCommit: ""
  GoVersion: go1.21.9
  Os: linux
  OsArch: linux/arm64
  Version: 5.0.2
```

### Podman in a container

No

### Privileged Or Rootless

Rootless

### Upstream Latest Release

Yes

### Additional environment details

On macOS, also using podman desktop.

```
23:13:02 $ podman version
Client:       Podman Engine
Version:      5.3.2
API Version:  5.3.2
Go Version:   go1.23.5
Git Commit:   85043bb1a3818102194afa82845cb63841067c9c
Built:        Tue Jan 21 10:41:34 2025
OS/Arch:      darwin/arm64

Server:       Podman Engine
Version:      5.0.2
API Version:  5.0.2
Go Version:   go1.21.9
Built:        Tue Apr 16 17:00:00 2024
OS/Arch:      linux/arm64
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by danishprakash**
We currently only pass parsed excludes from the ignore file to buildah and not the ignorefile path. This causes buildah to ignore `--ignorefile` if the specified file is empty.

Fixes #25147

### CODE REVIEW

File: pkg/api/handlers/compat/images_build.go [hunk 1]
```diff
@@ -289,6 +290,17 @@ func processBuildContext(query url.Values, r *http.Request, buildContext *BuildC
 		}
 	}
 
+	// Process dockerignore only if no explicit ignorefile was specified
+	if ignorefile := query.Get("ignorefile"); ignorefile != "" {
+		buildContext.IgnoreFile = filepath.Join(buildContext.ContextDirectory, ignorefile)
```
Comment (Honny1) [inline pkg/api/handlers/compat/images_build.go]: I think you must use `securejoin.SecureJoin`, because `filepath.Join` does not prevent directory traversal.

File: pkg/bindings/images/build.go [hunk 2]
```diff
@@ -188,6 +188,15 @@ func prepareParams(options types.BuildOptions) (url.Values, error) {
 		}
 		params.Set("excludes", bArgs)
 	}
+	if options.IgnoreFile != "" {
+		ignoreFile := options.IgnoreFile
+		if absIgnore, err := filepath.Abs(ignoreFile); err == nil {
+			if relPath, ok := strings.CutPrefix(absIgnore, options.ContextDirectory+string(filepath.Separator)); ok {
```
Comment (Honny1) [inline pkg/bindings/images/build.go:206]: What about ignore files outside of context? At least I would like to see tests for that.

**Review by Honny1 — CHANGES_REQUESTED** [review]

File: pkg/api/handlers/compat/images_build.go (hunk 1 above)
Comment (Honny1) [inline pkg/api/handlers/compat/images_build.go]: Also, don't forget about the local API. Will `validateLocalAPIPaths` catch a missing path?

File: pkg/bindings/images/build.go (hunk 2 above)
Comment (danishprakash) [reply pkg/bindings/images/build.go:206]: Thanks for the catch; this needed to be handled depending on whether the path was absolute or not; I've also handled this in the tests.

Comment (Honny1) [reply pkg/bindings/images/build.go:206]: This is still not resolved.

**Review by Honny1 — CHANGES_REQUESTED** [review]

File: pkg/bindings/images/build.go (hunk 2 above)
Comment (danishprakash) [reply pkg/bindings/images/build.go:206]: I've followed the same approach we use for handling containerfiles outside of the context dir, it's handled in `build()`. If the ignorefile is outside the context dir, the absolute path is forwarded to build, which then bundles the file at the absolute path in the tarball.

There's a test case that asserts this as well.

Comment (Honny1) [reply pkg/bindings/images/build.go:206]: I would need to check that.

Comment (Honny1) [reply pkg/bindings/images/build.go:206]: I run this command on my client system:
``` bash
$ pwd
/Users/jrodak/projects/containers/podman/test-build

$ ../bin/darwin/podman -c fedora-mac-vm build -t b-test-2 -f Containerfile --ignorefile ~/.dockerignore .
Key Passphrase:
Error: creating build executor: open /Users/jrodak/.dockerignore: no such file or directory

$ l
total 24
drwxr-xr-x@  5 jrodak  staff   160B Apr 29 14:38 .
drwxr-xr-x  81 jrodak  staff   2.5K Apr 29 14:36 ..
-rw-r--r--@  1 jrodak  staff     9B Apr 29 14:37 .dockerignore
-rw-r--r--@  1 jrodak  staff    56B Apr 29 14:38 Containerfile
-rw-r--r--@  1 jrodak  staff     5B Apr 29 14:37 testfile

$ ls ~/.dockerignore # Empty file
/Users/jrodak/.dockerignore

$ cat Containerfile
FROM fedora:latest
COPY . /testdir/
RUN find /testdir/

$ cat ./.dockerignore # docker ignore in context
testfile
```
> Important: Both Client and Server used changes from this PR, and they didn't share filesystems.

Additionally, when I changed the destination of the ignore file, it worked:
```bash

$  ls ../.dockerignore
../.dockerignore

$ ../bin/darwin/podman -c fedora-mac-vm build -t b-test-2 -f Containerfile --ignorefile ../.dockerignore . 
```

Comment (danishprakash) [reply pkg/bindings/images/build.go:206]: thanks for the catch; even though I followed what we do for containerfiles, it seems there's a fallback in buildah that results in success on different hosts (fs) for containerfiles. But this is better fixed in the server if the path is absolute and doesn't exist on disk (it's a client-side path that doesn't exist on the host).

Though I'm a bit concerned about how to test this? I added a couple of unit tests with the recent fix but they were superfluous, and I'm not sure how we can test this reliably with the CI.

**Comment by psss** [comment]
> tmt tests failed for commit [07eb88e](https://github.com/containers/podman/commit/07eb88e343893d4bb20760a8d98f418a331d95dd). @lsm5, @psss, @thrix please check.

Should be fixed once rebased on `main`.

File: pkg/bindings/images/build.go (hunk 2 above)
Comment (Honny1) [reply pkg/bindings/images/build.go:206]: This is hard to test because, for simplicity, tests are usually run on a single system. To test this properly, I set up a VM to act as the server.

Comment (danishprakash) [reply pkg/bindings/images/build.go:206]: I understand. Though if we can set up at least one environment that could catch situations like this, it'd be of great value. Please note I've not done any research on how this could be done yet, I'd like to though.

Comment (Honny1) [reply pkg/bindings/images/build.go:206]: Maybe it would be possible. WDYT? @timcoding1988 

~~Also, my test isn't throwing an error, but it seems the empty ignore file is being bypassed in favor of the ignore file in the context directory.~~

Comment (Honny1) [reply pkg/bindings/images/build.go:206]: Update: I used the wrong client binary. It works, but I found another issue: when `--ignorefile` points to a file outside the build context directory, the client adds it to the tarball using its full absolute host path, causing a spurious directory tree to appear in the built image.

Comment (danishprakash) [reply pkg/bindings/images/build.go:206]: Another good catch, though I believe that's existing behavior with how we create the tarball. I can reproduce this with Containerfiles outside the context directory. We could try a couple of things. First and foremost would be to simply tar the basename and send that from the client, but that has an obvious issue of file conflicts if the same file happens to be both inside and outside the context directory.

But I would ideally prefer if we create a new issue to discuss this further. Wdyt?

File: pkg/api/handlers/compat/images_build.go [hunk 3]
```diff
@@ -80,6 +81,7 @@ type BuildQuery struct {
 	IDMappingOptions        string             `schema:"idmappingoptions"`
 	IdentityLabel           bool               `schema:"identitylabel"`
 	Ignore                  bool               `schema:"ignore"`
+	IgnoreFile              string             `schema:"ignorefile"`
```
Comment (Honny1) [inline pkg/api/handlers/compat/images_build.go:88]: Please update the Swagger docs.

File: pkg/api/handlers/compat/images_build.go [hunk 4]
```diff
@@ -291,6 +299,33 @@ func processBuildContext(query url.Values, r *http.Request, buildContext *BuildC
...
+	// Process dockerignore only if no explicit ignorefile was specified
+	if ignorefile := query.Get("ignorefile"); ignorefile != "" {
+		if filepath.IsAbs(ignorefile) {
+			ignoreFilePath := filepath.Clean(ignorefile)
+			// In remote builds, the client may send an absolute path that exists
+			// only on the client filesystem. The file is bundled in the tarball
+			// and unpacked under the context directory at the same absolute path,
+			// so fall back to looking it up there if the absolute path is missing.
+			if err := fileutils.Exists(ignoreFilePath); err != nil {
+				ignoreFilePath = filepath.Join(buildContext.ContextDirectory, ignoreFilePath)
```
Comment (Honny1) [inline pkg/api/handlers/compat/images_build.go]: Maybe `SecureJoin`. Also, will it work with localAPI for MacOS and Windows?

File: pkg/bindings/images/build.go (hunk 2 above)
Comment (Honny1) [reply pkg/bindings/images/build.go:206]: Yep, new issue is good.

**Review by Honny1 — COMMENTED** [review]
Sorry for the late review. Can you please rebase on upstream main? Thanks.

File: pkg/api/handlers/compat/images_build.go (hunk 4 above)
Comment (danishprakash) [reply pkg/api/handlers/compat/images_build.go]: Okay, so this PR predates localAPI, so thanks for the catch, I've added a block handling ignorefile (same as containerfile) so it should cover localAPI

File: pkg/bindings/images/build.go (hunk 2 above)
Comment (danishprakash) [reply pkg/bindings/images/build.go:206]: Created https://github.com/podman-container-tools/podman/issues/29194; btw, this PR isn't blocked on that issue, that's a generic fix that will apply to ignorefile as well if and when it lands. So this PR can go in independently.

**Review by Honny1 — APPROVED** [review]
LGTM

PTAL @podman-container-tools/podman-reviewers @podman-container-tools/podman-maintainers

**Comment by mheon** [comment]
LGTM
